src/IO/AzureBlobStorage/copyAzureBlobStorageFile.cpp

```diff
@@ -0,0 +1,331 @@
+            task.block_id = getRandomASCIIString(64);
+            Azure::Core::IO::MemoryBodyStream memory(reinterpret_cast<const uint8_t *>(task.data), task.size);
+            block_blob_client.StageBlock(task.block_id, memory);
+
+            return task.block_id;
+        }
+
+
+        void waitForAllBackgroundTasks()
+        {
+            if (!schedule)
+                return;
+
+            std::unique_lock lock(bg_tasks_mutex);
+            /// Suppress warnings because bg_tasks_mutex is actually hold, but tsa annotations do not understand std::unique_lock
+            bg_tasks_condvar.wait(lock, [this]() {return TSA_SUPPRESS_WARNING_FOR_READ(num_added_bg_tasks) == TSA_SUPPRESS_WARNING_FOR_READ(num_finished_bg_tasks); });
+
+            auto & tasks = TSA_SUPPRESS_WARNING_FOR_WRITE(bg_tasks);
+            for (auto & task : tasks)
+            {
+                if (task.exception)
+                    std::rethrow_exception(task.exception);
+                block_ids.emplace_back(task.block_id);
+            }
+        }
+    };
+}
+
+
+void copyDataToAzureBlobStorageFile(
+    const std::function<std::unique_ptr<SeekableReadBuffer>()> & create_read_buffer,
+    size_t offset,
+    size_t size,
+    std::shared_ptr<Azure::Storage::Blobs::BlobContainerClient> & dest_client,
+    const String & dest_bucket,
+    const String & dest_key,
+    std::shared_ptr<AzureObjectStorageSettings> settings,
+    const std::optional<std::map<String, String>> & object_metadata,
+    ThreadPoolCallbackRunner<void> schedule,
+    bool for_disk_azure_blob_storage)
+{
+    UploadHelper helper{create_read_buffer, dest_client, offset, size, dest_bucket, dest_key, settings, object_metadata, schedule, for_disk_azure_blob_storage, &Poco::Logger::get("copyDataToAzureBlobStorageFile")};
+    helper.performCopy();
+}
+
+
+void copyAzureBlobStorageFile(
+    std::shared_ptr<Azure::Storage::Blobs::BlobContainerClient> src_client,
+    std::shared_ptr<Azure::Storage::Blobs::BlobContainerClient> dest_client,
+    const String & src_bucket,
+    const String & src_key,
+    size_t offset,
+    size_t size,
+    const String & dest_bucket,
+    const String & dest_key,
+    std::shared_ptr<AzureObjectStorageSettings> settings,
+    const ReadSettings & read_settings,
+    const std::optional<std::map<String, String>> & object_metadata,
+    ThreadPoolCallbackRunner<void> schedule,
+    bool for_disk_azure_blob_storage)
+{
+
+    if (size < max_single_operation_copy_size)
+    {
+        ProfileEvents::increment(ProfileEvents::AzureCopyObject);
+        if (for_disk_azure_blob_storage)
+            ProfileEvents::increment(ProfileEvents::DiskAzureCopyObject);
+        auto block_blob_client_src = src_client->GetBlockBlobClient(src_key);
+        auto block_blob_client_dest = dest_client->GetBlockBlobClient(dest_key);
+        auto uri = block_blob_client_src.GetUrl();
+        block_blob_client_dest.CopyFromUri(uri);
```

Comment:
There is another function `block_blob_client_dest.StartCopyFromUri()` which is a bit harder to use, however `StartCopyFromUri()` doesn't have the 256 MB limitation. I think we should use it.

Reply:
Why do you think `CopyFromUri` can be faster for smaller files? I haven't found any information about that. I think it can be faster for very small files, let's say less than 1MB (and not 256 MB) because of using less count of requests-responses to the Azure API (because when using `StartCopyFromUri` you have to make another request to understand if the copying is finished or not).

Reply:
yes, because of using less count of requests-responses to the Azure API

Reply:
I have added StartCopyFromUri under the flag 'use_native_copy' similar to S3. So we can check the performance & use native copy or multipart upload & choose what to use as well. Let me know if you have any other suggestions